### PR TITLE
 Feature #111717: API based tj-dashboard component creation

### PR DESCRIPTION
**Pull request by Renuka-S**

### CODE REVIEW

**Comment by coolbung** [comment]
The morris etc files should be placed in the respective widget folders not in the common asset folder.

File: tjdashboard/code/administrator/sql/uninstall.mysql.utf8.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+DROP TABLE IF EXISTS `#__tj_dashboard`;
```
Comment (coolbung) [inline tjdashboard/code/administrator/sql/uninstall.mysql.utf8.sql]: Remove this. Dont drop any tables on uninstall.

File: tjdashboard/code/administrator/sql/update.mysql.utf8.sql [hunk 2]
```diff
@@ -0,0 +1 @@
+/* Only premium users are allowed to update a component */
```
Comment (coolbung) [inline tjdashboard/code/administrator/sql/update.mysql.utf8.sql]: This file is not needed.

File: tjdashboard/code/administrator/tables/dashboards.php [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	 * @return  check
+	 *
+	 * @since  1.0.0
+	 */
+	public function check()
+	{
+		$db = JFactory::getDbo();
+
+		// If there is an ordering column and this is a new row then get the next ordering value
+		if (property_exists($this, 'ordering') && $this->dashboard_id == 0)
```
Comment (coolbung) [inline tjdashboard/code/administrator/tables/dashboards.php]: Are you supporting ordering for dashboards ?

File: tjdashboard/code/administrator/tables/dashboards.php [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+		}
+
+		$this->modified_by = JFactory::getUser()->id;
+		$this->modified_on = JFactory::getDate("now", "UTC")->tosql();
+
+		$this->alias = trim($this->alias);
+
+		if (empty($this->alias))
+		{
+			$this->alias = $this->title;
```
Comment (coolbung) [inline tjdashboard/code/administrator/tables/dashboards.php]: this is wrong. you should sluggify the title. Refer shika code.

File: tjdashboard/code/administrator/tjdashboard.xml [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    <description></description>
+    
+    
+    
+    <install> <!-- Runs on install -->
+        <sql>
+            <file driver="mysql" charset="utf8">sql/install.mysql.utf8.sql</file>
+        </sql>
+    </install>
+    <update> <!-- Runs on update -->
```
Comment (coolbung) [inline tjdashboard/code/administrator/tjdashboard.xml]: This is not how the update section works in the manifest.

File: tjdashboard/code/administrator/tjdashboard.xml [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+                <field name="save_history" default="0" />
+            </fieldset>
+        </fields>
+    </config>
+    <plugins>
+        <plugin group="search" plugin="tjdashboard"/>
+    </plugins>
+    
+</extension>
+<!-- Component built by the Joomla Component Creator -->
```
Comment (coolbung) [inline tjdashboard/code/administrator/tjdashboard.xml]: Remove all these comments

File: tjdashboard/code/administrator/tables/dashboards.php (hunk 3 above)
Comment (Renuka-S) [reply tjdashboard/code/administrator/tables/dashboards.php]: @coolbung Yes. For list view, we need order.

File: tjdashboard/code/site/assets/js/tjdashContentService.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+var tjdashContentService = {
+	postData: function(url, formData, params) {
+		if(!params){
+			params = {};
+		}
+
+		params['url']		= url;
+		params['data'] 		= formData;
+		params['type'] 		= typeof params['type'] != "undefined" ? params['type'] : 'POST';
+		params['async'] 	= typeof params['async'] != "undefined" ? params['async'] :false;
```
Comment (coolbung) [inline tjdashboard/code/site/assets/js/tjdashContentService.js]: why is this needed ?

File: tjdashboard/code/site/assets/js/tjdashContentUI.js [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+						data =  r.data;
+					}
+				}
+			);
+
+		return data;
+	}
+
+tjdashContentUI.widget = tjdashContentUI.widget ? tjdashContentUI.widget : {};
+tjdashContentUI.widget.apiurl = 'index.php?option=com_api&app=tjdashboard&resource=widget&format=raw&id=';
```
Comment (coolbung) [inline tjdashboard/code/site/assets/js/tjdashContentUI.js]: Move all URLs to class properties at the top of the class

File: tjdashboard/code/site/assets/js/tjdashContentUI.js [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+		 //console.log("library present");
+		 tjdashContentUI.renderer[library].init(method, data);
+	}
+	else
+	{
+		jQuery('<div class="alert alert-info">Library not found</div>').appendTo('".'+data['element']+'"');
+	}
+}
+
+tjdashContentUI.renderer.morris = tjdashContentUI.renderer.morris ? tjdashContentUI.renderer.morris : {};
```
Comment (coolbung) [inline tjdashboard/code/site/assets/js/tjdashContentUI.js]: Why is there morris code here ?

File: tjdashboard/code/site/assets/js/tjdashboard.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
+var tjdashboard=
```
Comment (coolbung) [inline tjdashboard/code/site/assets/js/tjdashboard.js]: what is this file for ?

File: tjdashboard/code/site/helpers/tjdashboard.php [hunk 11]
```diff
@@ -0,0 +1,92 @@
+<?php
```
Comment (coolbung) [inline tjdashboard/code/site/helpers/tjdashboard.php]: Why helper ? This should be in the widget model

File: tjdashboard/code/site/includes/tjdashboard.php [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+	/**
+	 * Retrieves a table from the table folder
+	 *
+	 * @param   mixed  $id  The Joomla user id
+	 *
+	 * @return	Mixed
+	 *
+	 * @since 	1.0
+	 **/
+	public static function getUser($id = null)
```
Comment (coolbung) [inline tjdashboard/code/site/includes/tjdashboard.php]: ??

File: tjdashboard/code/site/libraries/dashboard.php [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+			return $this->load($model->getState("dashboard_id"));
+		}
+	}
+
+	/**
+	 * get the dashboard widget Data
+	 *
+	 * @param   INT  $dashboardId  dashboard Id
+	 *
+	 * @return	Array
```
Comment (coolbung) [inline tjdashboard/code/site/libraries/dashboard.php]: Is this the right return type ?

File: tjdashboard/code/site/libraries/dashboards.php [hunk 14]
```diff
@@ -0,0 +1,30 @@
+<?php
```
Comment (coolbung) [inline tjdashboard/code/site/libraries/dashboards.php]: Why is this file needed ?

File: tjdashboard/code/site/models/dashboards.php [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+	 * @return  JDatabaseQuery
+	 *
+	 * @since   1.0
+	 */
+	protected function getListQuery()
+	{
+		$app = JFactory::getApplication();
+		$jinput = $app->input;
+
+		$dashboardId = $jinput->get('id', 0, 'INT');
```
Comment (coolbung) [inline tjdashboard/code/site/models/dashboards.php]: No getting input variables in model.

File: tjdashboard/code/administrator/tables/dashboards.php (hunk 4 above)
Comment (Renuka-S) [reply tjdashboard/code/administrator/tables/dashboards.php]: @coolbung Done in the code below.Checked for Shika code as well they have done same

File: tjdashboard/code/site/assets/js/tjdashContentService.js (hunk 7 above)
Comment (Renuka-S) [reply tjdashboard/code/site/assets/js/tjdashContentService.js]: Done

File: tjdashboard/code/site/helpers/tjdashboard.php (hunk 11 above)
Comment (Renuka-S) [reply tjdashboard/code/site/helpers/tjdashboard.php]: We can access this from anywhere where we already have widget id.

File: tjdashboard/code/site/libraries/dashboard.php (hunk 13 above)
Comment (Renuka-S) [reply tjdashboard/code/site/libraries/dashboard.php]: Done

File: tjdashboard/code/site/libraries/dashboards.php (hunk 14 above)
Comment (Renuka-S) [reply tjdashboard/code/site/libraries/dashboards.php]: Done

File: tjdashboard/code/administrator/sql/install.mysql.utf8.sql [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+
+`ordering` INT(11)  NOT NULL ,
+`state` TINYINT(1)  NOT NULL ,
+`checked_out` INT(11)  NOT NULL ,
+`checked_out_time` DATETIME NOT NULL ,
+`created_by` INT(11)  NOT NULL ,
+`modified_by` INT(11)  NOT NULL ,
+`title` VARCHAR(255)  NOT NULL ,
+PRIMARY KEY (`id`)
+) ENGINE=MyISAM DEFAULT COLLATE=utf8_general_ci;
```
Comment (coolbung) [inline tjdashboard/code/administrator/sql/install.mysql.utf8.sql]: When creating the install / update SQL, don't put the auto increment, engine and table collation. These should be picked up from database level defaults

File: tjdashboard/code/administrator/tjdashboard.xml [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    <creationDate>2017-09-15</creationDate>
+    <copyright>2017 Techjoomla</copyright>
+    <license>GNU General Public License version 2 or later; see LICENSE.txt</license>
+    <author>Techjoomla</author>
+    <authorEmail>extensions@techjoomla.com</authorEmail>
+    <authorUrl>http://techjoomla.com</authorUrl>
+    <version>CVS: 1.0.0</version>
+    <description></description>
+
+    <install> <!-- Runs on install -->
```
Comment (coolbung) [inline tjdashboard/code/administrator/tjdashboard.xml]: Where is the updates section ?

File: tjdashboard/code/site/libraries/dashboard.php [hunk 18]
```diff
@@ -0,0 +1,196 @@
+<?php
+/**
+ * @package     Joomla.Site
+ * @subpackage  com_truck
```
Comment (coolbung) [inline tjdashboard/code/site/libraries/dashboard.php]: truck ?? Ensure that the file header lines are correct. 

package should be Com_Tjdashboard
you can skip subpackage

File: tjdashboard/code/site/models/dashboards.php [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+	 * @return  JDatabaseQuery
+	 *
+	 * @since   1.0
+	 */
+	protected function getListQuery()
+	{
+		$app = JFactory::getApplication();
+
+		// Initialize variables.
+		$db    = JFactory::getDbo();
```
Comment (coolbung) [inline tjdashboard/code/site/models/dashboards.php]: - You should use the database object that is already provided by the model
- Also support a search filter to search by dashboards by name

File: tjdashboard/code/site/models/forms/dashboard.xml [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+			description="JGLOBAL_FIELD_ID_DESC"/>
+
+		<field
+			name="asset_id"
+			type="hidden"
+		/>
+
+		<field
+			name="title" type="text"
+			filter="safehtml"
```
Comment (coolbung) [inline tjdashboard/code/site/models/forms/dashboard.xml]: Does the title need to have HTML ?

File: tjdashboard/code/site/models/forms/dashboard.xml [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+			label="COM_TJDASHBOARD_FORM_LBL_DASHBOARD_TITLE"
+			description="COM_TJDASHBOARD_FORM_DESC_DASHBOARD_TITLE"
+			size="10" />
+
+		<field
+			name="alias" type="text"
```
Comment (coolbung) [inline tjdashboard/code/site/models/forms/dashboard.xml]: This should also change to string maybe

File: tjdashboard/code/site/models/forms/dashboard.xml [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+			filter="safehtml"
+			label="COM_TJDASHBOARD_FORM_LBL_DASHBOARD_ALIAS"
+			description="COM_TJDASHBOARD_FORM_DESC_DASHBOARD_ALIAS"
+			size="10" />
+
+		<field
```
Comment (coolbung) [inline tjdashboard/code/site/models/forms/dashboard.xml]: this should have a safehtml filter

File: tjdashboard/code/site/models/widgets.php [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+			$query->where($db->quoteName('size') . ' = ' . $db->escape($size));
+		}
+
+		// Add the list ordering clause.
+		$orderCol  = $this->getState('list.ordering');
+		$orderDirn = $this->getState('list.direction');
+
+		if (empty($orderCol))
+		{
+			$orderCol  = "dashboard_widget_id";
```
Comment (coolbung) [inline tjdashboard/code/site/models/widgets.php]: Shouldn't the default ordering be by ordering column ??

File: tjdashboard/code/site/router.php [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+			$view = $query['view'];
+
+			unset($query['view']);
+		}
+
+		if (isset($query['id']))
+		{
+			if ($view !== null)
+			{
+				$segments[] = $query['id'];
```
Comment (coolbung) [inline tjdashboard/code/site/router.php]: you should convert the id to the alias when building.. otherwise what's the point of having alias field ??

File: tjdashboard/code/site/router.php [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+	 * @return  array  The URL arguments to use to assemble the subsequent URL.
+	 *
+	 * @since   3.3
+	 */
+	public function build(&$query)
+	{
+		$segments = array();
+		$view     = null;
+
+		if (isset($query['task']))
```
Comment (coolbung) [inline tjdashboard/code/site/router.php]: This does not seem correct.. how are tasks handled when building and parsing ?

I suggest that for task SEF URLs you simply convert it to `/action/[task]`

File: tjdashboard/code/site/views/widgets/tmpl/default.xml [hunk 26]
```diff
@@ -0,0 +1,24 @@
...
+                        <![CDATA[COM_TJDASHBOARD_TITLE_ITEM_VIEW_WIDGET_DESC]]>
+        </message>
+    </layout>
+    <fields name="params">
+        <fieldset
+            name="basic"
+            label="COM_TJDASHBOARD_FIELDSET_ITEM_ID_SELECT_LABEL">
+            <field
+                name="dashboard_id"
+                query="SELECT `dashboard_id`, `title` FROM #__tj_dashboards ORDER BY `dashboard_id`"
```
Comment (coolbung) [inline tjdashboard/code/site/views/widgets/tmpl/default.xml]: Shouldn't you show only published dashboards ? Also, what about ordering ?

File: tjdashboard/plugins/api/tjdashboard/tjdashboard.php [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+		parent::__construct($subject, $config = array());
+
+		// Set resource path
+		ApiResource::addIncludePath(dirname(__FILE__) . '/tjdashboard');
+
+		// Load language files
+		$lang = JFactory::getLanguage();
+		$lang->load('plg_api_tjdashboard', JPATH_ADMINISTRATOR, '', true);
+
+		// Set the resource to be public
```
Comment (coolbung) [inline tjdashboard/plugins/api/tjdashboard/tjdashboard.php]: Why are all the resources public ? The read APIs can be public, but write APIs cannot be public.

File: tjdashboard/plugins/api/tjdashboard/tjdashboard/widget.php [hunk 28]
```diff
@@ -0,0 +1,75 @@
...
+
+/**
+ * Tjdashboard API widgets class
+ *
+ * @since  1.0.0
+ */
+class TjdashboardApiResourceWidget extends ApiResource
+{
+	/**
+	 * Function get for widgets record.
```
Comment (coolbung) [inline tjdashboard/plugins/api/tjdashboard/tjdashboard/widget.php]: Can you make the comments more meaningful

File: tjdashboard/code/administrator/sql/install.mysql.utf8.sql (hunk 16 above)
Comment (ttplsudhir) [reply tjdashboard/code/administrator/sql/install.mysql.utf8.sql]: @irfanhanfi123  @coolbung Agree with engine and table collation but why auto increment should not be there ?

File: tjdashboard/code/site/models/forms/dashboard.xml (hunk 21 above)
Comment (ttplsudhir) [reply tjdashboard/code/site/models/forms/dashboard.xml]: @coolbung  Why we should use filters for alias/title.User should able to add whatever they want we need to take care at point of rendering that we should escaping the html tags. We can look at the Article xml file for reference. What you think ?

File: tjdashboard/code/site/models/forms/dashboard.xml (hunk 22 above)
Comment (ttplsudhir) [reply tjdashboard/code/site/models/forms/dashboard.xml]: Done

File: tjdashboard/code/site/models/forms/dashboard.xml (hunk 20 above)
Comment (ttplsudhir) [reply tjdashboard/code/site/models/forms/dashboard.xml]: Removed

File: tjdashboard/code/site/models/dashboards.php (hunk 19 above)
Comment (ttplsudhir) [reply tjdashboard/code/site/models/dashboards.php]: Done ($this)

File: tjdashboard/code/site/libraries/dashboard.php (hunk 18 above)
Comment (ttplsudhir) [reply tjdashboard/code/site/libraries/dashboard.php]: Done

File: tjdashboard/code/site/models/widgets.php (hunk 23 above)
Comment (ttplsudhir) [reply tjdashboard/code/site/models/widgets.php]: Yeah Agree, Done

File: tjdashboard/code/site/models/dashboards.php [hunk 29]
```diff
@@ -52,10 +52,8 @@ public function __construct($config = array())
 	 */
 	protected function getListQuery()
 	{
-		$app = JFactory::getApplication();
-
 		// Initialize variables.
-		$db    = JFactory::getDbo();
+		$db    = $this->getDbo();
```
Comment (coolbung) [inline tjdashboard/code/site/models/dashboards.php]: Can you just use `$this->_db` ?

File: tjdashboard/code/site/models/forms/dashboard.xml (hunk 21 above)
Comment (coolbung) [reply tjdashboard/code/site/models/forms/dashboard.xml]: Ok for title. But why do you need HTML in the alias field ?

File: tjdashboard/code/site/views/widgets/tmpl/default.xml [hunk 30]
```diff
@@ -11,7 +11,7 @@
             label="COM_TJDASHBOARD_FIELDSET_ITEM_ID_SELECT_LABEL">
             <field
                 name="dashboard_id"
-                query="SELECT `dashboard_id`, `title` FROM #__tj_dashboards ORDER BY `dashboard_id`"
+                query="SELECT `dashboard_id`, `title` FROM #__tj_dashboards ORDER BY `ordering` where `state`=1"
```
Comment (coolbung) [inline tjdashboard/code/site/views/widgets/tmpl/default.xml]: Over here, I think ordering by `title asc` or `created date desc` might be better. What do you think ?

File: tjdashboard/code/administrator/sql/install.mysql.utf8.sql [hunk 31]
```diff
@@ -0,0 +1,15 @@
+CREATE TABLE IF NOT EXISTS `#__tj_dashboard` (
```
Comment (coolbung) [inline tjdashboard/code/administrator/sql/install.mysql.utf8.sql]: Can you create this SQL fresh.. why is there no alias column here ??

Also, where is the widgets table ?

File: tjdashboard/code/site/models/forms/dashboard.xml (hunk 21 above)
Comment (ttplsudhir) [reply tjdashboard/code/site/models/forms/dashboard.xml]: OK, Yeah alias should be string only as it will be in urls so there should not be tags there.

**Comment by coolbung** [comment]
Also.. the morris etc files should be in the respective plugins

File: tjdashboard/code/site/models/dashboards.php (hunk 29 above)
Comment (ttplsudhir) [reply tjdashboard/code/site/models/dashboards.php]: No, It is protected variable and we can use protected variable in only to it's child class not to the child's child class.

File: tjdashboard/code/site/views/widgets/tmpl/default.xml (hunk 30 above)
Comment (ttplsudhir) [reply tjdashboard/code/site/views/widgets/tmpl/default.xml]: Here I see , We are not even showing title of dashboard as the query selecting only dashboard id If we are going to show dashboards title there we should add title in select. 

In my opinion ordering column make sense, If you check with categories and article case they shows categories in article creation and it shows according to ordering column If you change order of categories it reflect in article form as well.

File: tjdashboard/code/administrator/tjdashboard.xml (hunk 17 above)
Comment (ttplsudhir) [reply tjdashboard/code/administrator/tjdashboard.xml]: @coolbung Little bit confused , Update about sql(tables)  or updateserver ?
1. To update the tables structure
     <update> 
         <schemas>
	       <schemapath type="mysql">sql/updates/mysql</schemapath>
	</schemas>
     </update>
2. To upgrade the 
<updateservers>
    <server type="extension" priority="1" name="Extension Update Site">http://example.com/extension.xml</server>
    </updateservers>

File: tjdashboard/code/administrator/sql/install.mysql.utf8.sql (hunk 31 above)
Comment (ttplsudhir) [reply tjdashboard/code/administrator/sql/install.mysql.utf8.sql]: Done

File: tjdashboard/code/administrator/tables/dashboards.php (hunk 3 above)
Comment (coolbung) [reply tjdashboard/code/administrator/tables/dashboards.php]: OK

File: tjdashboard/code/administrator/tables/dashboards.php (hunk 4 above)
Comment (coolbung) [reply tjdashboard/code/administrator/tables/dashboards.php]: OK

File: tjdashboard/code/site/helpers/tjdashboard.php (hunk 11 above)
Comment (coolbung) [reply tjdashboard/code/site/helpers/tjdashboard.php]: OK

File: tjdashboard/code/administrator/tjdashboard.xml (hunk 17 above)
Comment (coolbung) [reply tjdashboard/code/administrator/tjdashboard.xml]: OK

File: tjdashboard/code/site/models/forms/dashboard.xml (hunk 21 above)
Comment (coolbung) [reply tjdashboard/code/site/models/forms/dashboard.xml]: OK

File: tjdashboard/code/site/models/forms/dashboard.xml (hunk 22 above)
Comment (coolbung) [reply tjdashboard/code/site/models/forms/dashboard.xml]: OK

File: tjdashboard/plugins/api/tjdashboard/tjdashboard.php (hunk 27 above)
Comment (ttplsudhir) [reply tjdashboard/plugins/api/tjdashboard/tjdashboard.php]: Done

File: tjdashboard/code/site/views/dashboard/view.html.php [hunk 32]
```diff
@@ -0,0 +1,48 @@
...
+	 * @return void
+	 *
+	 * @throws Exception
+	 */
+	public function display($tpl = null)
+	{
+		$app = JFactory::getApplication();
+		$currentMenuItem = $app->getMenu()->getActive();
+		$params = $currentMenuItem->params;
+		$this->dashboardId = $params->get('dashboard_id', 0, 'INT');
```
Comment (coolbung) [inline tjdashboard/code/site/views/dashboard/view.html.php]: If I don't create a link to a specific dashboard, the id will be in the URL, how is that handled ?

File: tjdashboard/code/administrator/sql/install.mysql.utf8.sql [hunk 33]
```diff
@@ -0,0 +1,37 @@
...
+  `renderer_plugin` varchar(255) NOT NULL,
+  `ordering` int(11) NOT NULL,
+  `size` int(11) NOT NULL,
+  `params` text NOT NULL,
+  `autorefresh` int(11) NOT NULL,
+  `state` int(11) NOT NULL,
+  `created_on` datetime NOT NULL,
+  `modified_by` int(11) NOT NULL,
+  `modified_on` datetime NOT NULL,
+  PRIMARY KEY (`dashboard_widget_id`)
```
[the remark this answers is not in the document]
Comment (ttplsudhir) [reply tjdashboard/code/administrator/sql/install.mysql.utf8.sql]: @coolbung lets freeze the columns and correct use of it then I will create table accordingly and will implement same here.

File: tjdashboard/code/site/views/dashboard/view.html.php (hunk 32 above)
Comment (ttplsudhir) [reply tjdashboard/code/site/views/dashboard/view.html.php]: @coolbung Yes agree wanted to see the demo of it so I could able to solve it , as well as the permission related things also wanted to see it here in html.

File: tjdashboard/code/administrator/tables/dashboards.php [hunk 34]
```diff
@@ -0,0 +1,99 @@
...
+
+		// If there is an ordering column and this is a new row then get the next ordering value
+		if (property_exists($this, 'ordering') && $this->dashboard_id == 0)
+		{
+			$this->ordering = self::getNextOrder();
+		}
+
+		if ($this->dashboard_id == 0)
+		{
+			$this->created_by = JFactory::getUser()->id;
```
[the remark this answers is not in the document]
Comment (ttplsudhir) [reply tjdashboard/code/administrator/tables/dashboards.php]: @coolbung How could be that case , If some one else is owned dashboard then admin will be modifying that will not be creating it. Correct me if I assumed it wrong.
